Messages::GPUConnectionToWebProcess::CreateGraphicsContextGL/ReleaseGraphicsContextGL should only be handled when UseGPUProcessForWebGLEnabled is true

squelart opened the pull request:
#### cb9b0517ea6e2e1b9772f0929b29f445f1a64248
<pre>
Messages::GPUConnectionToWebProcess::CreateGraphicsContextGL/ReleaseGraphicsContextGL should only be handled when UseGPUProcessForWebGLEnabled is true
<a href="https://bugs.webkit.org/show_bug.cgi?id=276355">https://bugs.webkit.org/show_bug.cgi?id=276355</a>
<a href="https://rdar.apple.com/107941430">rdar://107941430</a>

Reviewed by Dan Glastonbury.

GPUProcessPreferencesForWebProcess::isWebGLEnabled is now
dependant on UseGPUProcessForWebGLEnabled, so that in case that
the GPU process is not to be used for WebGL, messages from the
GPU process related to WebGL will simply be ignored.

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):

Canonical link: <a href="https://commits.webkit.org/280843@main">https://commits.webkit.org/280843@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7247a6addf1160cd63dd622e04331aa9baa2856d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61419 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8242 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8430 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46828 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5848 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59827 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34820 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49946 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27657 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31589 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7246 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53540 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63102 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1711 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54050 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1717 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49957 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54168 "Found 3 new API test failures: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12788 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1454 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32954 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34040 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35124 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -138,7 +138,7 @@ class GPUConnectionToWebProcess
 
     bool isWebGPUEnabled() const { return m_preferences.isWebGPUEnabled; }
-    bool isWebGLEnabled() const { return m_preferences.isWebGLEnabled; }
+    bool isWebGLInGPUProcessEnabled() const { return m_preferences.isWebGLInGPUProcessEnabled; }
```
`InGPUProcess` seems redundant. This is a method on `GPUConnectionToWebProcess`, after all.

If we are going to rename this method, `isWebGPUEnabled` and `isWebXREnabled` should also be updated for consistency.

squelart replied:
I'm all for consistency, but I think it doesn't apply in this case:
- `isWebGPUEnabled()` and `isWebXREnabled()` are single boolean values reflecting the same flags in other processes.
- `isWebGLInGPUProcessEnabled()` is different, in that it combines the "traditional" app-wide `isWebGLEnabled()` with an additional flag allowing (or not) WebGL to run in the GPU process. I think the distinction here is important, and useful to highlight wherever it is used.
Please let me know if you can think of a better wording. `isGPUWebGLEnabled()`? `isWebGLEnabledAndUsesGPUProcess()`?

An alternative would be to have two separate & consistent flags `isWebGLEnabled()` and `useGPUProcessForWebGLEnabled()`, and check `isWebGLEnabled() && useGPUProcessForWebGLEnabled()` before handling GL messages from the GPU process.

squelart replied:
On reflection, I suppose that as you say we could interpret `GPUProcessPreferencesForWebProcess` members to be implicitly conditional upon each being allowed in the GPU process.
So `GPUProcessPreferencesForWebProcess::isWebGLEnabled()` means WebGL is currently enabled *and* it can run in the GPU process, etc.

So I'd be fine with reverting my name change; I just thought it was useful to add the clarification wherever it's used, to distinguish it at a glance from `WebPreferences::webGLEnabled()` (which does *not* include `useGPUProcessForWebGLEnabled`).

djg replied:
My other question disappeared, which was: When is useGPUProcessForWebGLEnabled ever false?

I don't think we support WebGL *not* in the GPU process. OpenGL is definitely blocked in WebContent and so might Metal. Do we support WebGL in WK1? (These are questions for @kkinnunen-apple)

squelart replied:
Alright, I've undone the renaming, it makes for a much simpler patch! 😄

squelart replied:
> When is useGPUProcessForWebGLEnabled ever false?

If it's never going to be false, should it just be removed to avoid confusion? 🤷

But in UnifiedWebPreferences.yaml, I see:
```
UseGPUProcessForWebGLEnabled: [...]
  defaultValue: [...]
    WebKit:
      "ENABLE(GPU_PROCESS_BY_DEFAULT) && ENABLE(GPU_PROCESS_WEBGL_BY_DEFAULT)": true
      "USE(GRAPHICS_LAYER_WC)": true
      default: false
```
So it could theoretically be false if some combination of these all-caps flags are falsy.

Anyway, in the meantime this flag exists, so just in case it *could* be set to false somehow, we may as well respect it where appropriate.

djg reviewed (approved):
:shipit:

squelart committed:
Messages::GPUConnectionToWebProcess::CreateGraphicsContextGL/ReleaseGraphicsContextGL should only be handled when UseGPUProcessForWebGLEnabled is true
https://bugs.webkit.org/show_bug.cgi?id=276355
rdar://107941430

Reviewed by Dan Glastonbury.

GPUProcessPreferencesForWebProcess::isWebGLEnabled is now
dependant on UseGPUProcessForWebGLEnabled, so that in case that
the GPU process is not to be used for WebGL, messages from the
GPU process related to WebGL will simply be ignored.

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):

Canonical link: https://commits.webkit.org/280843@main
